src/lib.rs

```diff
@@ -1011,6 +1011,144 @@ where
+    /// Removes the key (and its corresponding value) from this map.
+    /// This method does nothing if the key is not in the map.
+    /// Returns the previous value associated with the given key.
+    pub fn remove<'g>(&'g self, key: &K, guard: &'g Guard) -> Option<&'g V> {
+        let h = self.hash(key);
+
+        let mut table = self.table.load(Ordering::SeqCst, guard);
+
+        loop {
+            if table.is_null() {
+                break;
+            }
+
+            // safety: table is a valid pointer.
+            //
+            // we are in one of two cases:
+            //
+            //  1. if table is the one we read before the loop, then we read it while holding the
+            //     guard, so it won't be dropped until after we drop that guard b/c the drop logic
+            //     only queues a drop for the next epoch after removing the table.
+            //
+            //  2. if table is set by a Moved node (below) through help_transfer, it will use the
+            //     `next_table` raw pointer from inside the Moved. how do we know that that is safe?
+            //
+            //     we must demonstrate that if a Moved(t) is _read_, then t must still be valid.
+            //     FIXME cf. put
+            let t = unsafe { table.deref() };
+            let n = t.bins.len() as u64;
+            if n == 0 {
+                break;
+            }
+            let i = ((n - 1) & h) as usize;
+            let bin = t.bin(i, guard);
+            if bin.is_null() {
+                break;
+            }
+
+            // safety: bin is a valid pointer.
+            //
+            // there are two cases when a bin pointer is invalidated:
+            //
+            //  1. if the table was resized, bin is a move entry, and the resize has completed. in
+            //     that case, the table (and all its heads) will be dropped in the next epoch
+            //     following that.
+            //  2. if the table is being resized, bin may be swapped with a move entry. the old bin
+            //     will then be dropped in the following epoch after that happens.
+            //
+            // in both cases, we held the guard when we got the reference to the bin. if any such
+            // swap happened, it must have happened _after_ we read. since we did the read while
+            // pinning the epoch, the drop must happen in the _next_ epoch (i.e., the one that we
+            // are holding up by holding on to our guard).
+            match *unsafe { bin.deref() } {
+                BinEntry::Moved(next_table) => {
+                    table = self.help_transfer(table, next_table, guard);
+                }
+                BinEntry::Node(ref head) => {
+                    let head_lock = head.lock.lock();
+                    let mut old_val: Option<Shared<'g, V>> = None;
+                    let mut validated = false;
+
+                    if t.bin(i, guard) == bin {
+                        validated = true;
```

Comment:
I don't think the `validated` variable is actually necessary if you invert the clause above.